src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -0,0 +1,403 @@
+            if (_eventStream != IntPtr.Zero)
+            {
+                Interop.CoreServices.FSEventStreamStop(_eventStream);
+            }
+        }
+
+        private void WatchForFileSystemEventsThreadStart()
+        {
+            // Get this thread's RunLoop
+            _watcherRunLoop = Interop.CoreFoundation.CFRunLoopGetCurrent();
+            Debug.Assert(_watcherRunLoop != IntPtr.Zero);
+
+            // Schedule the EventStream to run on the thread's RunLoop
+            Interop.CoreServices.FSEventStreamScheduleWithRunLoop(_eventStream, _watcherRunLoop, Interop.CoreFoundation.kCFRunLoopDefaultMode);
+            if (Interop.CoreServices.FSEventStreamStart(_eventStream))
+            {
+                // Notify the StartRaisingEvents call that we are initialized and about to start
+                // so that it can return and avoid a race-condition around multiple threads calling Stop and Start
+                _runLoopStartedEvent.Set();
+
+                // Start the OS X RunLoop (a blocking call) that will pump file system changes into the callback function
+                Interop.CoreFoundation.CFRunLoopRun();
+
+                // When we get here, we've requested to stop so cleanup the EventStream and unschedule from the RunLoop
+                Interop.CoreServices.FSEventStreamStop(_eventStream);
+                Interop.CoreServices.FSEventStreamUnscheduleFromRunLoop(_eventStream, _watcherRunLoop, Interop.CoreFoundation.kCFRunLoopDefaultMode);
+            }
+            else
+            {
+                // An error occurred while trying to start the run loop so fail out
+                Interop.CoreServices.FSEventStreamUnscheduleFromRunLoop(_eventStream, _watcherRunLoop, Interop.CoreFoundation.kCFRunLoopDefaultMode);
+                OnError(new ErrorEventArgs(new IOException("", Marshal.GetLastWin32Error())));
+            }
+        }
+
+        private void FileSystemEventCallback( 
+            FSEventStreamRef streamRef, 
+            IntPtr clientCallBackInfo, 
+            size_t numEvents, 
+            String[] eventPaths, 
+            [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2)]
+            Interop.CoreServices.FSEventStreamEventFlags[] eventFlags,
+            [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2)]
+            FSEventStreamEventId[] eventIds)
+        {
+            Debug.Assert((numEvents.ToInt32() == eventPaths.Length) && (numEvents.ToInt32() == eventFlags.Length) && (numEvents.ToInt32() == eventIds.Length));
+
+            // Since renames come in pairs, when we find the first we need to search for the next one. Once we find it, we'll add it to this
+            // list so when the for-loop comes across it, we'll skip it since it's already been processed as part of the original of the pair.
+            System.Collections.Generic.List<FSEventStreamEventId> handledRenameEvents = new Collections.Generic.List<FSEventStreamEventId>();
```

Comment:
This is going to allocate a List even if there are no rename events.  It'd probably be better to initialize handledRenameEvents to null.  Then change the condition below that checks whether it contains something to be ```handledRenameEvents 1= null && handledRenameEvents.Contains(...)```, and preface the code that adds to the list with ```if (handledRenameEvents != null) { handledRenameEvents = new List<FSEventStreamEventId>();" so as to lazily initialize it.